README

Skytoop, an open source cloud desktop
====================================

I did Skytoop to learn Ajax, Web2Py and Jquery intensively.

How to use it ? 
---------------

* Start the server (python web2py.py)
* Go to http://localhost:8000/Skytoop/default/create_user to create a default user
* Now go to http://localhost:8000/Skytoop and login with model@skytoop.com, password = 123456

Functionnalities
----------------

* Filemanager (controllers/filemanager.py + views/filemanager)
** Multi file upload 
** Size limitation on total file uploaded (db/db_tables.py)
** Wallpapers can be set
* Public/Private Skytoop
* Widgets (static/widgets/)
** Widget Draw (static/widgets/drawZone) (the code is ugly)
** Widget Note (static/widgets/notes)
** Icons
* Ajax Search
* Widgets position synchronisation

Widgets
-------

The "Parent widget" (this is not a real inheritance, todo) is 
located in static/widgets/widgetsGen/jquery.widgetsGen.js

To add a new widget :

# Add the name (represented by his name in syncType) of the widget in :

── $.fn.widgetGen.select

The parent widget has a CRUD for child widgets :
── $.fn.widgetGen.save
── $.fn.widgetGen.update
── $.fn.widgetGen.remove
── $.fn.widgetGen.send

# Datas you can syncronize with the server are all fields begining 
with the prefix sync:


## code

$.fn.widgetGen.defaults = {
	syncType	: '',
	syncX		: 0,
	syncY		: 0,
	syncWidth	: 0,
	syncHeight	: 0,
	syncData1	: '',	 
	syncData2	: '',
	syncData3	: '',
	syncTitle	: '',
	syncId		: 0
    };

# For an example refer to static/widgets/notes/jquery.stickynote.js 
or static/widgets/widgetsGen/sample-widget.js

Files
-----

* JS for Desktop design are on static/desk/assets/javascript/*
* JS for the filemanager are on static/filemanager

### Controllers

controllers/
├── appadmin.py       # Appadmin
├── custom.py         # To change wallpaper and to switch on a public Skytoop
├── default.py        # Desktop main controllers
├── filemanager.py    # Filemanager
├── get_list.py       # On the index when you click on left bottom Skytoop c, get desktop list
├── icons.py          # CRUD for icons + favicon.ico grabbing
├── note_log.py       # On the index page for displaying recent changes
├── public.py	      # For search bar + public Skytoops 
├── vote_log.py	      # Vote log, for voting for features
└── widget.py         # CRUD widgets

### JS/CSS Files

static/
├── color-picker           # Color picker when customizing wallpaper
├── css			   # CSS Files
├── desk		   # Main folder for JS/CSS Desktop
│   └── assets
│       ├── images
│       │   ├── gui
│       │   ├── icons
│       │   └── misc
│       ├── javascripts    # @@@ Here are the most important files for Desktop
│       ├── scroll	   # Customization of the scroll bar
│       └── stylesheets	   # CSS files
├── filemanager		   # For the filemanager
│   ├── css
│   ├── images
│   │   └── fileicons
│   ├── jquery_contextmenu
│   │   └── images
│   ├── jquery_filetree
│   │   └── images
│   └── js
├── html5
├── images
├── js
├── multifile-upload	  # For the multiple file upload
├── pres-images
├── speed-bar		  # Images for the "speed bar" at the right of desktops
├── svg
├── tmp
├── vote_log		  # For the vote log
│   ├── css
│   └── js
├── wallpapers -> ../uploads/
└── widgets		  # Widgets
    ├── drawZone	  
    ├── notes
    │   ├── css
    │   └── images
    ├── template
    ├── widgetsGen
    └── winamp




CHANGELOG
*
*
*  Skytoop
*  by Strzelewicz Alexandre
*
*

8/03/10 :
- FEATURE ! :
* Wallpaper can be changed
* On Private + Public Skytoop !
* You can choose background color !
- Filemanager is heavy load protected

7/03/10 :
- Default desktop at the creation
- File upload problems fixed
- Lot of fixes
- Db pooling enhancements
- SQL queries optimized + cache
- GUI fixes
- Window enhancements
- Public Search bar ala Facebook
- Tutorial enhanced


28/02/10 :
- Notifications
- Added title to Zone drawing
- FEATURE - Zone link opener
* Open all links instantly !
- FEATURE - Search bar
- Skytoop searching enhanced
- Bug fixes

26/02/10 :
- FEATURE - SkyVote:
* You can now say us 
what idea you prefer !
- FEATURE - ShapeDesign :
* You can make icon package !
- Add documentation API
- Powerful element positionning

25/02/10 :

- Favicon 
- Icon positioning
- Url rewriting
- Filemanager remix
- Fixes
- Project cleanance
- Meta/header pages fixed
- Copyright added

Fixes/Bug/Enhancements :
- Better Icon position
- Urls rewriting
- Filemanager remix
- Favicon design
- Project clean




23/02/10 :
- Firefox style fixed
- Window resize fixed

22/02/10 :
- Postgresql
- Public Desktop in progress
- Client re-struct
- JQuery last versions
- Code exec speed optimized
- NEW FEATURE :
* Public Desktop !
* With nice URL
* Public Notes implemented
* Public Icons implemented
- Lot of fixes
- Easier navigation


21/02/10 :
- Better icon positioning
- Widget note enhanced
- Navigation enhancements

20/02/10 :
- Chercheur de skytoop 
- formulaire full ajax, tres clean
- virer framebreaker
- \n -> <br/> !!!!!!!!!!!!


19/02/10 :
- Icônes auto resize = compatible toutes resolution sans pert d'icones
- Ajout de speed shortcuts sur la droite du bureau
- % position icones
- note log sur la droite du screen
- correction partie admin + modif password page
- ajout de restriction a groupe
- autres

18/02/10 :

- fonctionnel et style sur firefox
- Modification des icones
- SMTP ok
- Postgresql Ok - pret au deploiement (manque juste modif sur filemanager.py)
- reorganisation des fichiers + cleaning
- fusion + minification des JS du filemanager = gain en perf
- correction par ci par la d'effet de style
- matrice sur le desktop en cours
- matrice testee et OK sur les notes (Math.round)

4/02/10 :
- amelioration de l'affichage du displaying des fichiers
- clear et blaisage du code du filemanager
- finaliser la navigation mac like dans l'explorer


30/01/10 :
- mettre le src d'une iframe que lors de son chargement
- amelioration visu filemanager
- smooth resize (iframe window iframe fix)
- redimensionnement automatique
- allegement du file manager
- MAJOR CHANGE : /!\ Ajout de l'upload multi fichier /!\

27/01/10 :
* Reprise en main du code & des web techniques
* Download ok dans le file manager
* File manager en iframe intégré


Features :

* Multi file upload

* Code is 15% load on server and 85% on client with the help of special Jquery addon created by me :p

Example :

        new JQD.winSys({
                    WindowTitle: 'Gen win #1',
                    IconId : {{=i.id}},
                    IconX : {{=i.x}},
                    IconY : {{=i.y}},
                    TemplateUrl : 'asd'
      }).buildWindow();

=== Will directly create a new window set on client side web browser !



models/db_tables.py
# -*- coding: utf-8 -*-
# by Strzelewicz Alexandre 


#########################################################
#
# Model / Database / Mailing / Logger
#
#########################################################

from gluon.tools import *
auth = Auth(globals(),db)               

auth.settings.hmac_key = 'sha512:21e79fb0-0623-4621-8943-37767690d266'
auth.define_tables(migrate=migrate)

auth.settings.mailer = mail
auth.settings.registration_requires_verification = True
auth.settings.registration_requires_approval = False

auth.messages.verify_email = 'Click on the link http://'+request.env.http_host+URL(r=request,c='default',f='user',args=['verify_email'])+'/%(key)s to verify your email'
auth.settings.reset_password_requires_verification = True
auth.messages.reset_password = 'Click on the link http://'+request.env.http_host+URL(r=request,c='default',f='user',args=['reset_password'])+'/%(key)s to reset your password'
auth.settings.create_user_groups = False

crud = Crud(globals(),db)
crud.settings.auth = None
service = Service(globals())                  
plugins = PluginManager()


#########################################################################
#
#
# Table definitions
#
#
#########################################################################

import datetime
now=datetime.datetime.now()
from gluon.tools import prettydate

if auth.is_logged_in():
    me=auth.user.id
else:
    me=None


#
# User setting
#
# Default 50mo max
#
db.define_table('auth_settings',
                # Data settings
                Field('stored_data_size', 'integer', default=0),
                # Max size to store
                Field('limit_stored_data_size', 'integer', default=50000000),
                # User settings
                Field('user_alias', 'string'),
                Field('public_actif', 'boolean', default=False),
                Field('user_id', db.auth_user, requires=IS_IN_DB(db, db.auth_user.id)))

#
# File manager
#
db.define_table('allfiles',
                Field('filename'),
                Field('filepath'),
                Field('parentpath'),
                Field('filetype'),
                Field('file','upload', autodelete=True),
                Field('content','text'),
                Field('datecreated','datetime',default=now),
                Field('datemodified','datetime',default=now),
                Field('filesize','integer'),
                Field('user_file',db.auth_user,default=me,requires=IS_IN_DB(db, db.auth_user.id)))

#########################################################
#
# Desktops
#
#########################################################

db.define_table('desktop',
                Field('desk_name', type='string', length=21, requires=IS_LENGTH(20,1)),
                Field('is_private', type='boolean', default=True),
                Field('is_default', type='boolean', default=True),
                Field('pwd', type='password', default=''),
                Field('wallpaper', length=256, default='/Skytoop/static/cloudy.png'),
                Field('back_color', length=10, default='#333', requires=IS_LENGTH(10,3)),
                Field('position', length=15, default='center center', requires=IS_LENGTH(15,4)))

db.define_table('entr_user_desktop',
                Field('user_link', db.auth_user),
                Field('desktop_link', db.desktop))

#########################################################
#
# Widgets
#
#########################################################

db.define_table('widgets',
                Field('type', type='string', length=21, requires=IS_LENGTH(20,1)),
                Field('x', type='integer'),
                Field('y', type='integer'),
                Field('width', type='integer'),
                Field('height', type='integer'),
                Field('title', type='string', length=128),
                Field('data1', type='string', length=256),
                Field('data2', type='string', length=256),
                Field('data3', type='string', length=256),
                Field('user_link', db.auth_user, notnull=True, requires=IS_NOT_EMPTY()),
                Field('desktop_link', db.desktop, notnull=True, requires=IS_NOT_EMPTY()))


db.define_table('icons',
                Field('title', requires=IS_NOT_EMPTY()),
                Field('x', type='integer'),
                Field('y', type='integer'),
                Field('direct_link', 'boolean'),
                Field('logo_link', type='string', length=128),
                Field('url', type='string', length=128),
                Field('user_link', db.auth_user, requires=IS_IN_DB(db, db.auth_user.id)),
                Field('desktop_link', db.desktop, requires=IS_IN_DB(db, db.desktop.id)))

#
# Shared widgets
#
db.define_table('widgets_entr',
                Field('type', type='string', length=21, requires=IS_LENGTH(20,1)),
                Field('x', type='integer'),
                Field('y', type='integer'),
                Field('width', type='integer'),
                Field('height', type='integer'),
                Field('title', type='string', length=128),
                Field('data1', type='string', length=256),
                Field('data2', type='string', length=256),
                Field('data3', type='string', length=256))

db.define_table('entr_desktop_widgets',
                Field('widget_link', db.widgets_entr),
                Field('desktop_link', db.desktop))

#########################################################
#
# Misc
#
#########################################################
#
# Note log book
#
db.define_table('note_log',
                Field('content', 'text'),
                Field('date_created', 'datetime', default=now))
#
# Vote log
#
db.define_table('vote_log',
                Field('content', 'text'),
                Field('content_full', 'text'),
                Field('author'),
                Field('date_created', 'datetime', default=now),
                Field('votes', 'integer', default=0))

#
# User who votes
# Many to many relation
#
db.define_table('entr_vote_user',
                Field('auth_user', db.auth_user),
                Field('vote_log', db.vote_log))
                


db.define_table('remarks',
                Field('content', type='text', requires=IS_NOT_EMPTY()),
                Field('author', db.auth_user))


imagetypes = set(['gif','jpg','jpeg','png','bmp', 'JPG', 'PNG', 'GIF', 'BMP'])
